tools/closed-by-commits

```diff
@@ -0,0 +1,361 @@
+
+            for issue in issues_dict.values():
+                results[issue].add(pr)
+
+        if not pr_nodes:
+            for issue_num in self._extract_issue_numbers(commit_message):
+                maybe_issue = self._fetch_issue(issue_num)
+                if maybe_issue is None:
+                    continue
+                results[maybe_issue].update()
+
+        return results
+
+    def analyze_range(self, base: str, head: str) -> list[Issue]:
+        console = Console(stderr=True)
+
+        repository = self.github.get_repo("zulip/zulip")
+        comparison = repository.compare(base, head)
+        commit_shas = [commit.sha for commit in comparison.commits]
+        console.log(f"Found {len(commit_shas)} commits")
+
+        issue_to_prs: dict[Issue, set[PullRequest]] = defaultdict(set)
+        with Progress(console=console) as progress:
+            task = progress.add_task("Processing commits...", total=len(commit_shas))
+            for sha in commit_shas:
+                progress.console.log(f"Processing {sha}")
+
+                for issue, prs in self.get_issues_for_commit(sha).items():
+                    issue_to_prs[issue].update(prs)
+                progress.advance(task)
+
+        for issue, prs in issue_to_prs.items():
+            for pr in prs:
+                issue.czo_urls.update(pr.czo_urls)
+
+            issue.closed_by_prs = sorted(prs, key=lambda p: p.number)
+
+        unique_prs = len({pr for prs in issue_to_prs.values() for pr in prs})
+        console.log(
+            f"Found {unique_prs} unique PRs closing {len(issue_to_prs)} issues",
+        )
+
+        return sorted(issue_to_prs.keys(), key=lambda x: x.number)
+
+
+def validate_github_token(value: str) -> str:
+    if not value.startswith("github_"):
+        raise typer.BadParameter("Github access tokens start with `github_`")
+    return value
+
+
+def main(
+    base_commit: str,
+    head_commit: str,
+    token: Annotated[
+        str,
+        typer.Option(
+            metavar="TOKEN",
+            envvar="GITHUB_TOKEN",
+            show_envvar=True,
+            callback=validate_github_token,
+            help=(
+                "Github access token; can be a fine-grained personal access token "
+                "with read-only access to 'Public repositories'.  "
+                "See https://github.com/settings/personal-access-tokens/new"
+            ),
+        ),
+    ],
+) -> int:
+    """Find issues which are closed in a commit range."""
+
+    lines = []
+    for issue in CommitRangeAnalyzer(token).analyze_range(base_commit, head_commit):
+        if not issue.closed_by_prs and not issue.czo_urls:
+            # Closed an issue that was never discussed anywhere?
+            continue
+
+        lines.append(f"## #{issue.number}: {issue.title}")
+        lines.extend(f" - **Closed by:** #{pr.number}: {pr.title}" for pr in issue.closed_by_prs)
+        lines.extend(f" - {czo_url}" for czo_url in sorted(issue.czo_urls))
```

Comment:
It'd be neat to turn these Zulip URLs into Markdown syntax for nice links. So e.g. instead of:
```md
 - https://chat.zulip.org/#narrow/channel/6-frontend/topic/Deleting.20dark-theme.2Ecss
```

it'd say:
```md
 - [#frontend > Deleting dark-theme.css](https://chat.zulip.org/#narrow/channel/6-frontend/topic/Deleting.20dark-theme.2Ecss)
```

which would make the target easier to read even in the raw output, and easier still after it's pasted into a Zulip message and rendered there.

No need for that to be in v1, though.

Reply:
Makes sense. Best to leave out and consider later, then.